pxtlib/tutorialValidator.ts

```diff
@@ -123,23 +125,37 @@ namespace pxt.tutorial {
     * Checks if the all required number of blocks for a tutorial step is used, returns a boolean
     * @param usersBlockUsed an array of strings
     * @param tutorialBlockUsed the next available index
-    * @return true if all the required tutorial blocks were used, false otherwise
+    * @param currRule the current rule with its TutorialRuleStatus
+    * @return a tutorial rule status for currRule
     */
-    function validateNumberOfBlocks(usersBlockUsed: pxt.Map<number>, tutorialBlockUsed: pxt.Map<number>): boolean {
+    function validateNumberOfBlocks(usersBlockUsed: pxt.Map<number>, tutorialBlockUsed: pxt.Map<number>, currRule: TutorialRuleStatus): TutorialRuleStatus {
         const userBlockKeys = Object.keys(usersBlockUsed);
-        const tutorialBlockKeys = Object.keys(tutorialBlockUsed);
+        let tutorialBlockKeys: string[] = []
+        if (tutorialBlockUsed != undefined) {
+            tutorialBlockKeys = Object.keys(tutorialBlockUsed);
+        }
+        let isValid: boolean = true;
+        let sArr: string[] = [];
+        sArr[0] = "These are the blocks you seem to be missing:";
+        let arrayIndex: number = 1;
         if (userBlockKeys.length < tutorialBlockKeys.length) { // user doesn't have enough blocks
-            return false;
+            isValid = false;
         }
         for (let i: number = 0; i < tutorialBlockKeys.length; i++) {
             let tutorialBlockKey = tutorialBlockKeys[i];
             if (!usersBlockUsed[tutorialBlockKey]) { // user did not use a specific block
-                return false;
-            }
-            if (usersBlockUsed[tutorialBlockKey] < tutorialBlockUsed[tutorialBlockKey]) { // user did not use enough of a certain block
-                return false;
+                sArr[arrayIndex] = "- " + tutorialBlockKey;
+                arrayIndex++;
```

Comment:
is there a reason to manually increment the index instead of using `sArr.push(item)` ?